fetcher.go

```diff
@@ -1,18 +1,159 @@
 package main
 
 import (
+	"bytes"
 	"context"
 	"fmt"
 	"io"
 	"net/http"
 	"net/url"
 	"path"
+	"strings"
+	"sync"
 
 	rulesspec "github.com/observatorium/api/rules"
+	"github.com/prometheus/prometheus/model/rulefmt"
+	"gopkg.in/yaml.v3"
 )
 
-type fetcher interface {
-	getRules(ctx context.Context) (rules io.ReadCloser, err error)
+// RulesObjstoreFetcher fetches rules for all configured tenants from the rules-objstore.
+type RulesObjstoreFetcher struct {
+	client     rulesspec.ClientInterface
+	tenants    []string
+	tenantsMtx sync.Mutex
+}
+
+// NewRulesObjstoreFetcher creates a new RulesObjtoreFetcher.
+// The tenants list must be deduplicated otherwise, rules groups will not be unique.
+func NewRulesObjstoreFetcher(baseURL string, tenants []string, client *http.Client) (*RulesObjstoreFetcher, error) {
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	baseURLParsed, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse RulesObjtoreFetcher URL: %w", err)
+	}
+
+	rulesClient, err := rulesspec.NewClient(baseURLParsed.String(), rulesspec.WithHTTPClient(client))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create rules-objstore client: %w", err)
+	}
+
+	return &RulesObjstoreFetcher{
+		client:  rulesClient,
+		tenants: tenants,
+	}, nil
+}
+
+type tenantFetchResult struct {
+	tenant string
+	res    *http.Response
+	err    error
+}
+
+// GetTenantsRules fetches rules for all configured tenants from the rules-objstore.
+func (f *RulesObjstoreFetcher) GetTenantsRules(ctx context.Context) (io.ReadCloser, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	concurrency := 10
+	var sem = make(chan struct{}, concurrency)
+	results := make(chan tenantFetchResult)
+
+	// Launch goroutines that fetch rules for each tenant concurrently.
+	go func() {
+		var wg sync.WaitGroup
+		defer close(results)
+		defer wg.Wait() // Wait for all goroutines to finish before closing results channel.
+
+		// tenants can be changed concurrently, we copy the list to avoid locking for too long.
+		f.tenantsMtx.Lock()
+		tenants := make([]string, len(f.tenants))
+		copy(tenants, f.tenants)
+		f.tenantsMtx.Unlock()
+
+		for _, tenantID := range f.tenants {
+			// Use semaphore to limit concurrency, and return early if context is cancelled.
+			select {
+			case <-ctx.Done():
+				results <- tenantFetchResult{tenantID, nil, ctx.Err()}
+				return
+			case sem <- struct{}{}:
+			}
+
+			// Launch goroutine to fetch rules for a tenant.
+			wg.Add(1)
+			go func(tenantID string) {
+				defer wg.Done()
+				defer func() { <-sem }()
```

Comment:
Same note here about sequential deferred calls where order is very important and that can be collapsed into a single one.

Reply:
For this one, order does not matter. Grouping anyway.